app/bundles/CoreBundle/Entity/UpsertTrait.php

```diff
@@ -1,37 +0,0 @@
-<?php
-
-declare(strict_types=1);
-
-namespace Mautic\CoreBundle\Entity;
-
-trait UpsertTrait
```

Comment:
Removing this will break the Custom Objects plugin. See https://github.com/mautic/mautic/pull/11746

Reply:
so, there is a file in Mautic only used for potential plugins? interesting :) 

Wouldn't it be more logical to make a `app/bundles/CoreBundle/Entity/UpsertBase.php` class that implements the interface? and that base class is extended in plugins?
Or is here is a reason this is a trait?

Reply:
> so, there is a file in Mautic only used for potential plugins? interesting :)

It's more complicated than that. It exists in Acquia's fork of Mautic and serves another purpose in the core. However, this change was not pushed to Mautic OSS yet.

The Custom Objects plugin was released as OSS last year. The community started using it. Then we pushed some update that broke it for the community and so we quickly patched it with adding this trait to the next Mautic version so we'd fix it. I have GH Actions CI in progress for the plugin to ensure it won't happen, but there are other priorities.

So this was a quick fix and the intentional PR where it's used in the core is coming soon. It will all make sense in the end.

TBH I'm not clear on whether inheritance is less evil than traits. Both have pros and cons.